### DISCUSSION TITLE
Tiny-ALU Skeleton Restructuring

### DISCUSSION THREAD

**Discussion by RonZolti** [pyuvm/pyuvm#141] [General]
PYUVM Verification

April 15, 2023

Following the excellent skeleton provided by Ray Salemi and others in https://github.com/pyuvm/pyuvm , for handling a robust chip verification under Python, I'd like to suggest a small contribution which enables an arbitrary expansion of the verification platform.

The original Tiny-ALU example existing in the git-hub consists of an 'engine' (which drives the verification process) and 'DUT-like' portion which implements the specific activities carried-out by the emulated unit. The boundaries between the two entities are somewhat violated.

In a restructuring process, I modified the Tiny-ALU example such that it has been separated from the 'engine' portion - which may now be used as a general framework for testing an arbitrary type of DUT (or numerous of DUTs, simultaneously) – as required – per the system under test.

The current architecture yields identical results to the original framework.

I attach here the modified engine structure py files (but with docx extension - due to site's regulations). Side files are not added here. 
Arbitrary DUT modules handling may be readily added

For additional details, please contact me: [ronzolti@gmail.com](mailto:ronzolti@gmail.com)

[testbench.docx](https://github.com/pyuvm/pyuvm/files/11240416/testbench.docx)
[BfmConcentrator.docx](https://github.com/pyuvm/pyuvm/files/11240417/BfmConcentrator.docx)

**Comment by raysalemi** [comment]
Hi Ron,

Word cannot open these files. 

I recommend creating a new project on GitHub that uses **pyuvm** as a prerequisite.

**Comment by RonZolti** [comment]
Thank you Ray

My mistake. It was a raw 'py' file. These files may be renamed to extension 'py' and opened using any text editor.

Alternatively, I attach here a 'docx' version of these files.
[testbench.docx](https://github.com/pyuvm/pyuvm/files/11275245/testbench.docx)
[BfmConcentrator.docx](https://github.com/pyuvm/pyuvm/files/11275246/BfmConcentrator.docx)

**Comment by RonZolti** [comment]
Hello Ray

Thank you for notifying me.
I've reopened the discussion, under 'general' sction:

PYUVM Verification #143

with a proper file format.

I sent them here anyway, in py & docx files' format.
